ifmeorg/ifme
# Require recaptcha after 3 failed logins

# Description

This PR adds a recaptcha requirement if a user has failed to login 3 times. Failed attempts are tracked by Devise on a per user basis, so 3 failed logins with the same email will trigger the recaptcha requirement but 3 failed logins with different usernames will not.

## More Details

This PR adds the [ambethia/recaptcha](https://github.com/ambethia/recaptcha) gem for recaptchas. This gem was suggested in both the issue and in the [Devise Wiki](https://github.com/heartcombo/devise/wiki/How-To:-Use-Recaptcha-with-Devise). The implementation is based off of the Devise [example for logins](https://github.com/heartcombo/devise/wiki/How-To:-Use-Recaptcha-with-Devise#devisesessionscontroller). However, additional logic was required since we only want to show the recaptcha after 3 failed logins. A cookie is given to users who needed a recaptcha after their last failed login in case the user leaves the site before logging in and then comes back. Additionally, the recaptcha only display/is required if the recaptcha site key/secret key are set. This should ease local development.

This PR uses recaptcha V2, which requires the user to click an "I am not a robot" check box or solve a "challenge" by selecting images that contain a certain item. The frequency with which the image challenges appear can be adjusted within the recaptcha admin console.

Note:  We will need to set the site key and secret key ENV variables in prod. This is my first PR here, so I could probably use some help. This change could still be merged without them, we just won't show/require the recaptcha until then.

If you are reviewing this PR, leave a comment here and I will grant you access to the admin console for our recaptcha. Once you have been granted access, you can go to https://www.google.com/recaptcha/admin to view the keys.

## Validation Steps
It's up to you if you actually want to run through these steps. I'm not positive what the process for reviewing PRs consists of, but here's how you can check out the changes locally...

Once given access to recaptcha site/secret keys (comment and I'll add you)...
1. In `config/env/development.env` Set `RECAPTCHA_SITE_KEY` and `RECAPTCHA_SECRET_KEY` ENV variables
2. Try to sign in as a user with an incorrect password 3 times
3. Verify that after the 3rd failed attempt a recaptcha will appear
4. Before signing in, navigate away from the page, then come back to the sign in page. The recaptcha should still be there since you have a cookie set
5. Try to sign in with a different email. Since this other email doesn't require a recaptcha, the cookie will be cleared. Go to the sign in page and try to sign in with the original email that requires a recaptcha. The recaptcha field will not be present since the last user you attempted to sign in with did not require it, but if you try to sign in as the original user again, you'll be prompted to fill in a recaptcha
6. Ensure you can't submit the sign in form until the recaptcha has been completed
7. Sign in successfully

Might be a good idea to validate that the recaptcha doesn't appear/isn't required if the RECAPTCHA_SITE_KEY/RECAPTCHA_SECRET_KEY ENV variables aren't present


## Corresponding Issue

https://github.com/ifmeorg/ifme/issues/1921

# Screenshots

<img width="500" alt="Screen Shot 2020-11-27 at 6 02 37 PM" src="https://user-images.githubusercontent.com/32438079/100547351-51faa100-3234-11eb-9bdc-48a66afa9b88.png">

<img width="500" alt="recaptcha-challenge" src="https://user-images.githubusercontent.com/32438079/100547322-2b3c6a80-3234-11eb-94b0-0bd31cf8523b.png">

![recaptcha-example](https://user-images.githubusercontent.com/32438079/100547340-414a2b00-3234-11eb-949b-499baee564db.gif)


---

Reviewing this pull request? Check out our [Code Review Practices](https://github.com/ifmeorg/ifme/wiki/Code-Review-Practices) guide if you haven't already!
